data/tutorials/language/1ms_01_functors.md

```diff
@@ -371,7 +371,7 @@ let _ =
 
 Check the program's behaviour using `opam exec -- dune exec funkt < dune`.
 
-**Note**: The functor `IterPrint.Make` returns a module that exposes the type from the injected dependency (here first `List.t` then `Array.t`). That's why a `with type` constraint is needed. When parametrising other something not exposed by the module (and _implementation detail_), the `with type` constraint is not needed.
+**Note**: The functor `IterPrint.Make` exposes the type from the injected dependency (here first `List.t` then `Array.t`) as the type `t` required of the result module. That's why a `with type` constraint is needed. When customising a type that's not exposed by the result module (i.e. an _implementation detail_), the `with type` constraint is not needed. As a further consequence of this, we could even define a local type `t` within the `struct` block for use within the implementation, and such a type `t` would not shadow the type `t` of the result module.
```

Comment:
Thanks, @neuroevolutus, that note needs to be clarified. Your text is better than is initial one, do you think this improves on it:

```suggestion
**Note**: Modules received and returned by `IterPrint.Make` both have a type `t`. The `with type ... :=` constraint is needed to make the two `t` identical. This allows functions from the injected dependency and result module to use the same type. When the parameter's contained type is not exposed by the result module (i.e. when it is an _implementation detail_), the `with type` constraint is unnecessary. 
```

Regarding your last sentence:

> As a further consequence, we could even define a local type `t` within the `struct` block for use within the implementation, and such a type `t` would not shadow the type `t` of the result module.

This is a bit hard to grasp because it requires 1/ figuring out what the code would look like if modified with a local `t` and 2/ figuring out the shadowing concern.

However, I believe you have a valid point. This is a matter that needs to be addressed. Could we address it somewhere else in the document or do we need an additional example? We may even need a complete section to address naming and shadowing concerns in relationship with functor definitions.

What is our opinion on this?

Reply:
Thanks so much for the rewording. I think it's a lot more succinct and even clearer now. I made sure to commit your suggestion.

> This is a bit hard to grasp because it requires 1/ figuring out what the code would look like if modified with a local t and 2/ figuring out the shadowing concern.

I agree; that is a good point :sweat_smile:. I think adding a new section with a concrete example to address these concerns would be really nice. I'll work on adding this section and make a new commit to get your feedback.

Reply:
As a small aside, I added a small section on empty variants to the data types tutorial to give context for the use of a singular `|` as a placeholder type definition.